localstack/http/websocket.py

```diff
@@ -0,0 +1,326 @@
+import functools
+import typing as t
+
+from werkzeug import Response
+from werkzeug._internal import _wsgi_decoding_dance
+from werkzeug.datastructures import EnvironHeaders, Headers
+from werkzeug.sansio.request import Request as _SansIORequest
+from werkzeug.wsgi import _get_server
+
+from .asgi import ASGIWebSocket, WebSocketEnvironment, WebSocketListener
+
+
+class WebSocketError(IOError):
+    """Base class for websocket errors"""
+
+    pass
+
+
+class WebSocketDisconnectedError(WebSocketError):
+    """Raised when the client has disconnected while the server is still trying to receive data."""
+
+    default_code = 1005
+    """https://asgi.readthedocs.io/en/latest/specs/www.html#disconnect-receive-event-ws"""
+
+    def __init__(self, code: int = None):
+        self.code = code if code is not None else self.default_code
+        super().__init__(f"Websocket disconnected code={self.code}")
+
+
+class WebSocketProtocolError(WebSocketError):
+    """Raised if there is a problem in the interaction between app and the websocket server."""
+
+    pass
+
+
+class WebSocket:
+    """
+    High-level interface to interact with a websocket after a handshake has been completed with
+    `WebsocketRequest.accept()`.
+    """
+
+    request: "WebSocketRequest"
+    socket: ASGIWebSocket
+    encoding: str = "utf-8"
+
+    def __init__(self, request: "WebSocketRequest", socket: ASGIWebSocket):
+        self.request = request
+        self.socket = socket
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    def __iter__(self):
+        while True:
+            try:
+                yield self.receive()
+            except WebSocketDisconnectedError:
+                break
+
+    def send(self, text_or_bytes: str | bytes, timeout: float = None):
+        """
+        Send data to the websocket connection.
+
+        :param text_or_bytes: the data to send. Use strings for text-mode sockets (default).
+        :param timeout: the timeout in seconds to wait before raising a timeout error
+        """
+        if isinstance(text_or_bytes, str):
+            self.socket.send(
+                {
+                    "type": "websocket.send",
+                    "bytes": None,
+                    "text": text_or_bytes,
+                },
+                timeout=timeout,
+            )
+        else:
+            self.socket.send(
+                {
+                    "type": "websocket.send",
+                    "bytes": text_or_bytes,
+                    "text": None,
+                },
+                timeout=timeout,
+            )
+
+    def receive(self) -> str:
```

Comment:
q: The receive method here always returns a string. How would I implement a binary communication on the websocket (like some kind of compressed protocol)?

Reply:
i can see that ASGI handles binary/text mode per-frame automatically based on a flag set in the frame. this is also the way client libraries do this. this means that `receive` can basically return `str | bytes`, which is what i implemented now. there's also a test for this